### PR TITLE
Use dynamic casting for handoffs

### PR DESCRIPTION
**Pull request by MingweiSamuel**
Instead of using shared ownership of handoffs, have the `Hydroflow` struct own the handoffs and cast them as needed for the operators.

---

Description of how everything currently works:

1. When you create a subgraph, the input/output ports store a `Rc<Cell<Option<HandoffId>>>` to later identify which handoffs they have been hooked up via. The other half of the `Rc` is stored internally in a `Subgraph` (runnable) closure.
2. When an `OutputPort` is connected to an `InputPort` the handoff `H` is created and stored and the new `HandoffId` is given to both the ports so that the corresponding `Rc<Cell<...>>`s in the internal closures know where the handoff is.
3. When the `Subgraph` closure is run, it grabs out the handoff IDs, looks them up in the `Hydroflow.handoffs` vec, and downcast_ref's them into the correct `H` type and makes the `Recv`/`SendCtx`s. Then the actual subgraph work code can run

---

Current clunkyness:
- (1) Input/OutputPort still use a similar `Rc<Cell<Option<HandoffId>>>` mechanism (instead of `Rc<RefCell<H>>`) to coordinate handoff Ids.
- (2) `TeeingHandoff` has been temporarily commented out.
- (3) Now requires GATs because the `HandoffList` has to deal with references :/
- (4) `Subraph` runnable takes in a `&[HandoffData]` to do the indexing/casting which is MESSY.

Future work TODO:
- (1) A separate layer of indirection in which actual input and output points on each operator are identified may be better as the current setup prevents serialization and thread-crossing.
- (2/3/4) Add a proper state context API to replace `&[HandoffData]`
  - State context API should allow richer handoff functionality to reimplement `TeeingHandoff`.

### CODE REVIEW

File: hydroflow/src/scheduled/handoff/mod.rs [hunk 1]
```diff
@@ -18,7 +16,9 @@ pub trait CanReceive<T> {
 /**
  * A handle onto the metadata part of a [Handoff], with no element type.
  */
-pub trait HandoffMeta {
+pub trait HandoffMeta: Any {
+    fn any_ref(&self) -> &dyn Any;
```
Comment (justinj) [inline hydroflow/src/scheduled/handoff/mod.rs:20]: How come this can't be done at the call site?

File: hydroflow/src/scheduled/mod.rs [hunk 2]
```diff
@@ -329,45 +376,52 @@ impl Hydroflow {
...
+        let handoff = H::default();
         self.handoffs.push(HandoffData::new(
-            output_port.handoff.clone(),
+            handoff,
             output_port.op_id,
             input_port.op_id,
         ));
 
-        // Add successor.
+        // Add successor & predacessor.
```
Comment (justinj) [inline hydroflow/src/scheduled/mod.rs:422]: nit: `predecessor`

File: hydroflow/src/scheduled/mod.rs [hunk 3]
```diff
@@ -506,6 +562,44 @@ mod tests {
...
 
+    #[test]
+    fn test_basic_variadic() {
+        let mut df = Hydroflow::new();
+        let source_handle = df.add_source(move |send: &SendCtx<VecHandoff<usize>>| {
+            send.give(Some(5));
+        });
+        let sink_handle = df.add_sink(move |recv: &RecvCtx<VecHandoff<usize>>| {
+            for v in recv.take_inner().into_iter() {
+                assert_eq!(5, v);
```
Comment (justinj) [inline hydroflow/src/scheduled/mod.rs:573]: Can we also assert somewhere in here that this actually runs?

File: hydroflow/src/scheduled/mod.rs [hunk 4]
```diff
@@ -506,6 +562,44 @@ mod tests {
...
+            move |_: &[&RecvCtx<VecHandoff<usize>>], send: &[&SendCtx<VecHandoff<usize>>]| {
+                send[0].give(Some(5));
+            },
+        );
+        let (mut sink_handle, _) = df.add_n_in_m_out(
+            1,
+            0,
+            move |recv: &[&RecvCtx<VecHandoff<usize>>], _: &[&SendCtx<VecHandoff<usize>>]| {
+                for v in recv[0].take_inner().into_iter() {
+                    assert_eq!(5, v);
```
Comment (justinj) [inline hydroflow/src/scheduled/mod.rs:595]: ditto here

**Review by justinj — APPROVED** [review]
Remind me again of the motivation for this change? Is it so that the scheduler has access to the handoffs and can do things like run compaction on them?

File: hydroflow/src/scheduled/handoff/mod.rs (hunk 1 above)
Comment (MingweiSamuel) [reply hydroflow/src/scheduled/handoff/mod.rs:20]: It is weird but necessary unfortunately. Basically we have the concrete type `dyn HandoffMeta` but  [`downcast_ref`](https://doc.rust-lang.org/std/any/trait.Any.html#implementations) is on `dyn Any`. If we call it directly we get the impl on `dyn HandoffMeta`, so we either need [trait upcasting (unstable)](https://doc.rust-lang.org/nightly/unstable-book/language-features/trait-upcasting.html) or just have to manually implement this indirection. Longer version here: https://stackoverflow.com/a/33687996/2398020

**Comment by MingweiSamuel** [comment]
> Remind me again of the motivation for this change? Is it so that the scheduler has access to the handoffs and can do things like run compaction on them?

It removes the need for shared ownership of handoffs, as the overall `Hydroflow` instance now owns them. It also means in the future the handoff handles don't have to store RcRCs in them. And it also does let the scheduler/overall instance mess with the state if it ever needs to.

And it's also about 5% faster for most benchmarks.

**Pull request merged**